src/io/flutter/editor/InlinePreviewViewController.java

```diff
@@ -0,0 +1,403 @@
+    visibleRect = newRectangle;
+    if (getDescriptor() == null || getData().getMarker() == null) {
+      if (!visible) {
+        visible = true;
+        onVisibleChanged();
+      }
+      return;
+    }
+    final TextRange marker = getData().getMarker();
+    if (marker == null) return;
+
+    final Point start = offsetToPoint(marker.getStartOffset());
+    final Point end = offsetToPoint(marker.getEndOffset());
+    final boolean nowVisible = newRectangle == null || newRectangle.y <= end.y && newRectangle.y + newRectangle.height >= start.y ||
+                               updateVisiblityLocked(newRectangle);
+    if (visible != nowVisible) {
+      visible = nowVisible;
+      onVisibleChanged();
+    }
+  }
+
+  @Override
+  public void dispose() {
+  }
+
+  @Override
+  protected Component getComponent() {
+    return getEditor().getComponent();
+  }
+
+  @Override
+  protected VirtualFile getVirtualFile() {
+    return getEditor().getVirtualFile();
+  }
+
+  @Override
+  protected Document getDocument() {
+    return getEditor().getDocument();
+  }
+
+  @Override
+  protected void showPopup(Point location, DiagnosticsNode node) {
+    location =
+      SwingUtilities.convertPoint(
+        getEditor().getContentComponent(),
+        location,
+        getEditor().getComponent()
+      );
+    popup = PropertyEditorPanel
+      .showPopup(data.context.inspectorGroupManagerService, getEditor(), node, node.getCreationLocation().getLocation(),
+                 data.context.flutterDartAnalysisService, location);
+  }
+
+  @Override
+  TextRange getActiveRange() {
+    return getData().getMarker();
+  }
+
+  @Override
+  void setCustomCursor(@Nullable Cursor cursor) {
+    if (getEditor() == null) {
+      // TODO(jacobr): customize the cursor when there is not an associated editor.
+      return;
+    }
+    getEditor().setCustomCursor(this, cursor);
+  }
+
+  // Determine zOrder of overlapping previews.
+  // Ideally we should work harder to prevent overlapping.
+  public int getPriority() {
+    int priority = 0;
+    if (popupActive()) {
+      priority += 20;
+    }
+    if (isVisiblityLocked()) {
+      priority += 10;
+    }
+
+    if (isSelected) {
+      priority += 5;
+    }
+
+    if (getDescriptor() != null) {
+      priority += 1;
+    }
+
+    if (screenshot == null && (elements == null || elements.isEmpty())) {
+      priority -= 5;
+      if (getDescriptor() != null) {
+        priority -= 100;
+      }
+    }
+    else {
+      if (hasCurrentHits() || _mouseInScreenshot) {
+        priority += 10;
+      }
+    }
+    if (_mouseInScreenshot) {
+      priority += 20;
+    }
+    return priority;
+  }
+
+  @Override
+  public void paint(@NotNull Editor editor, @NotNull RangeHighlighter highlighter, @NotNull Graphics g) {
+    if (editor != getEditor()) {
+      // Don't want to render on the wrong editor. This shouldn't happen.
+      return;
+    }
+    if (getEditor().isPurePaintingMode()) {
```

Comment:
Can getEditor() return a different value here? If not, just use editor.

Reply:
The value should never change. However, the rest of the code will assume getEditor() versus passing in the editor to the paint call so I think using getEditor() everywhere is reasonable.

Reply:
This weird api is due to supporting RangeHighlighter where the editor is typically passed in rather than tightly coupled to the object as it is in this case.